llvm/lib/RemoteCachingService/CAS/GRPCRelayCAS.cpp

```diff
@@ -0,0 +1,504 @@
+  Expected<CASID> parseID(StringRef ID) final;
+  Expected<ObjectRef> store(ArrayRef<ObjectRef> Refs,
+                               ArrayRef<char> Data) final;
+  CASID getID(ObjectRef Ref) const final;
+  CASID getID(ObjectHandle Handle) const final;
+  Optional<ObjectRef> getReference(const CASID &ID) const final;
+  ObjectRef getReference(ObjectHandle Handle) const final;
+  Expected<ObjectHandle> load(ObjectRef Ref) final;
+  Error validate(const CASID &ID) final {
+    // Not supported yet. Always return success.
+    return Error::success();
+  }
+  uint64_t getDataSize(ObjectHandle Node) const final;
+  Error forEachRef(ObjectHandle Node,
+                   function_ref<Error(ObjectRef)> Callback) const final;
+  ObjectRef readRef(ObjectHandle Node, size_t I) const final;
+  size_t getNumRefs(ObjectHandle Node) const final;
+  ArrayRef<char> getData(ObjectHandle Node,
+                         bool RequiresNullTerminator = false) const final;
+
+  // For sending file path through grpc.
+  Expected<ObjectRef>
+  storeFromOpenFileImpl(sys::fs::file_t FD,
+                        Optional<sys::fs::file_status> Status) override;
+
+  // API that uses the file API.
+  Expected<ObjectRef> storeFile(StringRef Content);
+  Expected<StringRef> loadFile(ObjectRef Node);
+
+private:
+  InMemoryIndexValueT &indexHash(ArrayRef<uint8_t> Hash) const {
+    assert(Hash.size() == ServiceHashSize && "unexpected hash size");
+    SmallVector<uint8_t, sizeof(InMemoryIndexT::HashT)> ExtendedHash(
+        sizeof(InMemoryIndexT::HashT));
+    llvm::copy(Hash, ExtendedHash.begin());
+    return *Index.insertLazy(ExtendedHash, [](auto ValueConstructor) {
+      ValueConstructor.emplace(nullptr);
+    });
+  }
+
+  CASID getID(const InMemoryIndexValueT &I) const {
+    ArrayRef<uint8_t> Hash = I.Data->getHash().take_front(ServiceHashSize);
+    return CASID::create(&getContext(), toStringRef(Hash));
+  }
+
+  const InMemoryCASData &asInMemoryCASData(ObjectHandle Ref) const {
+    uintptr_t P = Ref.getInternalRef(*this);
+    return *reinterpret_cast<const InMemoryCASData *>(P);
+  }
+
+  InMemoryIndexValueT &asInMemoryIndexValue(ObjectRef Ref) const {
+    uintptr_t P = Ref.getInternalRef(*this);
+    return *reinterpret_cast<InMemoryIndexValueT *>(P);
+  }
+
+  ObjectRef toReference(const InMemoryCASData &O) const {
+    return toReference(O.getIndex());
+  }
+
+  ObjectRef toReference(const InMemoryIndexValueT &I) const {
+    return makeObjectRef(reinterpret_cast<uintptr_t>(&I));
+  }
+
+  const InMemoryCASData &getInMemoryCASData(ObjectHandle OH) const {
+    return *reinterpret_cast<const InMemoryCASData *>(
+        (uintptr_t)OH.getInternalRef(*this));
+  }
+
+  ObjectHandle getObjectHandle(const InMemoryCASData &Node) const {
+    assert(!(reinterpret_cast<uintptr_t>(&Node) & 0x1ULL));
+    return makeObjectHandle(reinterpret_cast<uintptr_t>(&Node));
+  }
+
+  const InMemoryCASData &
+  storeObjectImpl(InMemoryIndexValueT &I,
```

Comment:
The reason for this is simply because the life time of object is the same as CAS, so CAS needs to keep the memory alive till it is destroyed.

But yes, we can change the design but that is a much bigger problem than implement a GRPC CAS.